tiny_dnn/core/framework/tensor.h

```diff
@@ -61,42 +61,41 @@
 
 namespace tiny_dnn {
 
-template<typename U = float_t>
+template<typename U = float_t, size_t kDimensions = 4>
 class Tensor {
 public:
     /*
      * Initializes an empty tensor.
      */
     Tensor()
     {
-        reshape(0, 0, 0, 0);
+        reshape(std::array<size_t, kDimensions>());
     }
 
     /*
      * Create a tensor of the given dimension.
-     * It is assumed that a tensor will hold data in NxWxHxD order,
-     * where:
-     *  N the batch axis
-     *  W the width axis
-     *  H the heigth axis
-     *  D the depth axis
+     * It is assumed that a tensor will hold data in C-style nD array, i.e
+     * row-major order:  the rightmost index “varies the fastest”.
      *
      *  Data will be hold by a std::vector with 64bytes alignment.
      */
-    explicit Tensor(const size_t d0,
-                    const size_t d1,
-                    const size_t d2,
-                    const size_t d3) {
+    /*explicit Tensor(const size_t d0) {
         reshape(d0, d1, d2, d3);
-    }
+    }*/ //TODO(Randl): variadic template version
 
-    explicit Tensor(const std::array<size_t, 4>& shape) {
-        reshape(shape[0], shape[1], shape[2], shape[3]);
+    explicit Tensor(const std::array<size_t, kDimensions>& shape) {
+        reshape(shape);
     }
 
     explicit Tensor(const std::vector<size_t>& shape) {
-        assert(shape.size() == 4);
-        reshape(shape[0], shape[1], shape[2], shape[3]);
+        reshape(shape);
+    }
+
+    explicit Tensor(std::initializer_list<size_t> const& shape)  {
+        assert(shape.size() == kDimensions);
```

Comment:
@edgarriba I think shape.size() isn't known at compile time. Not sure about it though. Gotta check.